src/diagnostic/GraphProbe.cpp

```diff
@@ -0,0 +1,420 @@
+/** Copyright 2017-2021 CNRS-AIST JRL and CNRS-UM LIRMM */
+
+#include <tvm/diagnostic/GraphProbe.h>
+
+#include <iomanip>
+#include <iostream>
+#include <queue>
+#include <string>
+
+namespace
+{
+using namespace tvm::diagnostic;
+using namespace tvm::graph::internal;
+
+std::string indent(int val) { return std::string(val, ' '); }
+
+// shared code for any and none (see below)
+template<bool b>
+bool check(const std::vector<GraphProbe::OutputVal> & outputVal, std::function<bool(const Eigen::MatrixXd &)> select)
+{
+  for(const auto & v : outputVal)
+  {
+    if(select(std::get<Eigen::MatrixXd>(v)))
+      return b;
+  }
+  return !b;
+}
+
+// any(outputVal, select) returns true if any value v in outputVal is such that select(v) is true, and false otherwise
+constexpr auto any = check<true>;
+// none(outputVal, select) returns true if no value v in outputVal is such that select(v) is true, and false otherwise
+constexpr auto none = check<false>;
+
+Eigen::MatrixXd transposeIfVector(const Eigen::MatrixXd & M)
+{
+  if(M.cols() == 1)
+    return M.transpose();
+  else
+    return M;
+}
+
+// find the input corresponding to the direct dependency
+template<typename InputContainer>
+const Log::Input & findInput(const InputContainer & s, const Log::DirectDependency & d)
+{
+  for(const auto & i : s)
+  {
+    if(i.owner == d.owner && i.id == d.input && i.source == d.source)
+    {
+      return i;
+    }
+  }
+  throw std::runtime_error("Input not found");
+}
+
+// find the input corresponding to the input dependency
+template<typename InputContainer>
+const Log::Input & findInput(const InputContainer & s, const Log::InputDependency & d)
+{
+  for(const auto & i : s)
+  {
+    if(i.owner == d.owner && i.id == d.input && i.source == d.source)
+    {
+      return i;
+    }
+  }
+  throw std::runtime_error("Input not found");
+}
```

Comment:
Same function so you can template the second parameter as well

Reply:
Done.